src/cryptonote_core/uptime_proof.cpp

```diff
@@ -0,0 +1,111 @@
+
+  crypto::generate_signature(hash, keys.pub, keys.key, sig);
+  crypto_sign_detached(sig_ed25519.data, NULL, reinterpret_cast<unsigned char *>(hash.data), sizeof(hash.data), keys.key_ed25519.data);
+}
+
+//Deserialize from a btencoded string into our Proof instance
+uptime_proof::Proof::Proof(const std::string& serialized_proof)
+{
+  try {
+    lokimq::bt_dict bt_proof = lokimq::bt_deserialize<lokimq::bt_dict>(serialized_proof);
+    //snode_version <X,X,X>
+    lokimq::bt_list& bt_version = var::get<lokimq::bt_list>(bt_proof["version"]);
+    int k = 0;
+    for (lokimq::bt_value const &i: bt_version){
+      snode_version[k++] = static_cast<uint16_t>(lokimq::get_int<unsigned>(i));
+    }
+    //timestamp
+    timestamp = lokimq::get_int<unsigned>(bt_proof["timestamp"]);
+    //public_ip
+    public_ip = lokimq::get_int<unsigned>(bt_proof["public_ip"]);
+    //storage_port
+    storage_port = static_cast<uint16_t>(lokimq::get_int<unsigned>(bt_proof["storage_port"]));
+    //pubkey_ed25519
+    pubkey_ed25519 = tools::make_from_guts<crypto::ed25519_public_key>(var::get<std::string>(bt_proof["pubkey_ed25519"]));
+    //pubkey
+    std::string pubkey_str = var::get<std::string>(bt_proof["pubkey_ed25519"]);
+    pubkey = (!pubkey_str.empty()) ? tools::make_from_guts<crypto::public_key>(pubkey_str) :tools::make_from_guts<crypto::public_key>(var::get<std::string>(bt_proof["pubkey_ed25519"]));
+    //qnet_port
+    qnet_port = lokimq::get_int<unsigned>(bt_proof["qnet_port"]);
+    //storage_lmq_port
+    storage_lmq_port = lokimq::get_int<unsigned>(bt_proof["storage_lmq_port"]);
+    //storage_version
+    lokimq::bt_list& bt_storage_version = var::get<lokimq::bt_list>(bt_proof["storage_version"]);
+    k = 0;
+    for (lokimq::bt_value const &i: bt_storage_version){
+      storage_version[k++] = static_cast<uint16_t>(lokimq::get_int<unsigned>(i));
+    }
+    //lokinet_version
+    lokimq::bt_list& bt_lokinet_version = var::get<lokimq::bt_list>(bt_proof["lokinet_version"]);
+    k = 0;
+    for (lokimq::bt_value const &i: bt_lokinet_version){
+      lokinet_version[k++] = static_cast<uint16_t>(lokimq::get_int<unsigned>(i));
+    }
+  } catch (const std::exception& e) {
+    std::cerr << "deserialization failed: " << e.what();
+  }
+}
+
+crypto::hash uptime_proof::hash_uptime_proof(const uptime_proof::Proof& proof)
+{
+  crypto::hash result;
+
+  std::string serialized_proof = lokimq::bt_serialize(uptime_proof::bt_encode_uptime_proof(proof));
+  size_t buf_size = serialized_proof.size();
+  crypto::cn_fast_hash(serialized_proof.data(), buf_size, result);
+  return result;
+}
+
+lokimq::bt_dict uptime_proof::bt_encode_uptime_proof(const uptime_proof::Proof& proof)
+{
+  return lokimq::bt_dict {
+    {"version", lokimq::bt_list{{proof.snode_version[0], proof.snode_version[1], proof.snode_version[2]}}},
+    {"pubkey",  (proof.pubkey == proof.pubkey_ed25519) ? "" : tools::view_guts(proof.pubkey)},
```

Comment:
If these are equal then we can omit the key entirely, rather than set it to an empty string.  You'll have to do something like:
```C++
lokimq::bt_dict proof{
   ... // all the fields *other* than pubkey
};
if (proof.pubkey != proof.pubkey_ed25519) proof["pubkey"] = tools::view_guts(proof.pubkey);
return proof;
```
and then when retrieving it:
```C++
if (auto it = bt_proof.find("proof"); it != bt_proof.end())
    pubkey = var::get<std::string>(*it);
else
    std::memcpy(pubkey.data, pubkey_ed25519.data, 32);
```